### PR TITLE
Add optional name parameter to buf mod init

### LINKED ISSUE #1648: Optionally allow module name in `buf mod init` command
It'd be nice to have `buf mod init` take an optional arg or flag for the module name. E.g.,

```
buf mod init buf.build/mfridman/api
```

Or 

```
buf mod init --name buf.build/mfridman/api
```

### PR DESCRIPTION
**Pull request by tonyli233**
Fixes #1648.

As response to [this issue](https://github.com/bufbuild/buf/issues/1648), I added a new arg for `buf mod init`. 

`buf mod init` now can take an input 
E.g. `buf mod init buf.build/bufbuild/buf`

### CODE REVIEW

File: private/buf/cmd/buf/command/mod/modinit/modinit.go
Comment (elliotmjackson) [inline private/buf/cmd/buf/command/mod/modinit/modinit.go:194]: have another go at this, try conform to the style-guide and describe var names

File: private/buf/cmd/buf/command/mod/modinit/modinit.go
Comment (elliotmjackson) [inline private/buf/cmd/buf/command/mod/modinit/modinit.go:63]: is this an "inputHashTag"? whats the definition?

File: private/buf/cmd/buf/command/mod/modinit/modinit.go
Comment (elliotmjackson) [inline private/buf/cmd/buf/command/mod/modinit/modinit.go:131]:
```suggestion
	var writeConfigOptions []bufconfig.WriteConfigOption
	name := container.Arg(1)
	if name != "" {
		writeConfigWithName, err := bufconfig.WriteConfigWithName(name)
		if err != nil {
			return err
		}
		writeConfigOptions = append(writeConfigOptions, writeConfigWithName)
	}
```

File: private/buf/cmd/buf/command/mod/modinit/modinit.go
Comment (elliotmjackson) [inline private/buf/cmd/buf/command/mod/modinit/modinit.go:123]: well done

File: private/bufpkg/bufconfig/bufconfig.go
Comment (marekbuild) [inline private/bufpkg/bufconfig/bufconfig.go:105]: Does this follow the style guide?

**Comment by marekbuild** [comment]
@tonyli233 @elliotmjackson Do we use `--name` anywhere else?

**Comment by elliotmjackson** [comment]
I'm unsure but we aren't using it here either... description is out of date, moving to draft

File: private/buf/cmd/buf/command/mod/modinit/modinit.go
Comment (elliotmjackson) [inline private/buf/cmd/buf/command/mod/modinit/modinit.go:46]: I'm not sure this counts as an "input" you will need to have a look at the precedent for `input` elsewhere in the codebase. This may be better off as `argument` or `arg`... I'm not sure, see what you can find

File: private/bufpkg/bufconfig/bufconfig.go
Comment (tonyli233) [reply private/bufpkg/bufconfig/bufconfig.go:105]: I think so. This conforms with the code surrounded. The parameter name is the module name passed in by user which has a key called ```name``` in the config file.

File: private/buf/cmd/buf/command/mod/modinit/modinit.go
Comment (tonyli233) [reply private/buf/cmd/buf/command/mod/modinit/modinit.go:46]: I checked the precedence of ```input``` in the codebase. Looks like ```input``` refers to files in other commands. I didn't find ```argument``` or ```arg``` elsewhere. I think ```module-name``` will be a better way to describe.

Comment (elliotmjackson) [reply private/buf/cmd/buf/command/mod/modinit/modinit.go:46]: [module-name](https://docs.buf.build/tour/push-a-module#module-names) looks appropriate.

File: private/bufpkg/bufconfig/bufconfig.go
Comment (tonyli233) [reply private/bufpkg/bufconfig/bufconfig.go:105]: I was trying to hide `moduleIdentity` in ```modinit.go``` since ´modinit.go´ is only setting the config with given parameters. 
I also agree that pulling the `bufmoduleref.ModuleIdentityForString(name)` call to `modinit.go` will make the functions uniform to other wrappers.

Comment (bufdev) [reply private/bufpkg/bufconfig/bufconfig.go:105]: All of our public types need documentation - looking at the style guide, this doesn't appear to be in there, so we should add

**Comment by bufdev** [comment]
This needs some thinking as to what we should do here - this is a one-way door, and adding a new argument to a command is the ultimate one-way door.

**Review by elliotmjackson — CHANGES_REQUESTED** [review]
blocking merge until @bufdev green light

File: private/bufpkg/bufconfig/bufconfig.go
Comment (tonyli233) [reply private/bufpkg/bufconfig/bufconfig.go]: The way of setting this config is changed. This function is no longer needed. 

I'll take a look at the style guide and keep in mind in future.

File: private/buf/cmd/buf/command/mod/modinit/modinit.go
Comment (bufdev) [inline private/buf/cmd/buf/command/mod/modinit/modinit.go:47]: @joshcarp do you like this?

**Review by bufdev — APPROVED** [review]
Approved pending @joshcarp's feedback

File: private/buf/cmd/buf/command/mod/modinit/modinit.go
Comment (joshcarp) [reply private/buf/cmd/buf/command/mod/modinit/modinit.go:47]: TBH we shouldn't introduce a new placeholder `<module-name>`. 
We should either:
- Use `[name]` because it's consistent with the `name` field in `buf.yaml` 
- Use `[module]` because we have a somewhat consistent idea of what a module is, and the context here implies that you'd want to add a name
- Use `[buf.build/owner/repository]` if we want to be the most user friendly
- The problem with this last one is that there's a lot of complexity on the users to understand the difference between `repository` and `module`. IMO the slightly better option would probably be to avoid using `repository` altogether in this case (because we're technically not "creating" a repository on the BSR, yet)

I think the option that will cause the least confusion would be
```
  buf mod init [buf.build/owner/foobar]
```
This:
- gives the optional form (with square brackets)
- avoids the ambiguity in what a `name` is
- shows the form as what is should be `buf.build/owner/foobar`
- avoids the mental strain of understanding what a `repository` is

```suggestion
		Use:   name + " [buf.build/owner/foobar]",
```

**Review by elliotmjackson — APPROVED** [review]

**Review by bufdev — APPROVED** [review]

**Review by joshcarp — APPROVED** [review]
A small nit, but overall LGTM

**Pull request merged**
